app/views/course/assessment/answer/programming/_test_cases.html.slim

```diff
@@ -13,41 +13,71 @@ table.table.table-striped.table-hover
   thead
     tr
       th = Course::Assessment::Question::ProgrammingTestCase.human_attribute_name(:identifier)
+      th = Course::Assessment::Question::ProgrammingTestCase.human_attribute_name(:expression)
+      th = Course::Assessment::Question::ProgrammingTestCase.human_attribute_name(:expected)
       th = Course::Assessment::Question::ProgrammingTestCase.human_attribute_name(:hint)
       th = t('.passed')
   tbody
     tr
-      th colspan=4
+      th colspan=5
         = t('.public')
     - public_test_cases.each do |test_case|
       - test_case_result = answer_test_results_hash[test_case]
       = content_tag_for(:tr, test_case) do
         th = format_html(test_case.identifier)
+        td = format_html(test_case.expression)
+        td = format_html(test_case.expected)
         td = format_html(test_case.hint) if !test_case_result || !test_case_result.passed?
         td
           - if test_case_result && test_case_result.passed?
             = fa_icon 'check'.freeze
-          - else
+          - elsif test_case_result
             = fa_icon 'times'.freeze
   - if can?(:grade, answer.answer)
     tbody
       tr
-        th colspan=4
+        th colspan=5
           = t('.private')
       - private_test_cases.each do |test_case|
+        - test_case_result = answer_test_results_hash[test_case]
         = content_tag_for(:tr, test_case) do
           th = format_html(test_case.identifier)
+          td = format_html(test_case.expression)
+          td = format_html(test_case.expected)
+          td = format_html(test_case.hint) if !test_case_result || !test_case_result.passed?
           td
+            - if test_case_result && test_case_result.passed?
+              = fa_icon 'check'.freeze
+            - elsif test_case_result
+              = fa_icon 'times'.freeze
+  - if can?(:grade, answer.answer)
```

Comment:
We should only display this when submission is submitted or graded. `can?(:grade, answer.answer) && !submission.attempting?`. And this can be merged with the if statement above ? so both of them need such check.